src/abstractdataframe/reshape.jl

```diff
@@ -185,39 +193,71 @@ long = stack(wide)
-    # `value` integer indicating which column has values
-    refkeycol = CategoricalArray{Union{eltype(df[rowkey]), Missing}}(df[rowkey])
-    # make sure we report only levels actually present in rowkey column
-    # this is consistent with how the other unstack method based on groupby works
+function unstack(df::AbstractDataFrame, rowkeys::Int, colkey::Int, value::Int)
+    refkeycol = copy(CategoricalArray{Union{eltype(df[rowkeys]), Missing}}(df[rowkeys]))
     droplevels!(refkeycol)
-    valuecol = df[value]
     keycol = CategoricalArray{Union{eltype(df[colkey]), Missing}}(df[colkey])
+    valuecol = df[value]
     Nrow = length(refkeycol.pool)
     Ncol = length(keycol.pool)
-    payload = DataFrame(Any[similar_missing(valuecol, Nrow) for i in 1:Ncol], map(Symbol, levels(keycol)))
-    nowarning = true
+    df2 = DataFrame(Any[similar_missing(valuecol, Nrow) for i in 1:Ncol], map(Symbol, levels(keycol)))
+    _unstack(df, rowkeys, colkey, value, keycol, valuecol, df2, refkeycol)
+end
+
+function _unstack(df::AbstractDataFrame, rowkeys::Int,
+                  colkey::Int, value::Int, keycol, valuecol, df2, refkeycol)
+    Nrow = length(refkeycol.pool)
+    Ncol = length(keycol.pool)
+    hadmissing = false # have we encounered missing in refkeycol
+    mask_filled = falses(Ncol, Nrow) # has a given [col,row] entry been filled?
+    nowarning = true # do we print duplicate entries warning
+    nowarning2 = true # do we print missing in keycol
+    keycol_pool = Vector{Int}(CategoricalArrays.order(keycol.pool))
+    refkeycol_pool = Vector{Int}(CategoricalArrays.order(refkeycol.pool))
     for k in 1:nrow(df)
-        j = Int(CategoricalArrays.order(keycol.pool)[keycol.refs[k]])
-        i = Int(CategoricalArrays.order(refkeycol.pool)[refkeycol.refs[k]])
-        if i > 0 && j > 0
-            if nowarning && !ismissing(payload[j][i])
-                warn("Duplicate entries in unstack.")
-                nowarning = false
+        keycol_refs = keycol.refs[k]
+        if keycol_refs == 0 # we have found missing in colkey
+            if nowarning2
+              warn("Missing value in colkey variable at row $k. Skipping.")
+              nowarning2 = false
+            end
+            continue # skip processing it
+        end
+        j = keycol_pool[keycol_refs]
+        refkeycol_refs = refkeycol.refs[k]
+        if refkeycol_refs == 0 # we have found missing in rowkeys
+            if !hadmissing # if it is the first time we have to add a new row
+                hadmissing = true
+                # we use the fact that missing is greater than anything
+                push!(df2, missings(ncol(df2)))
+                mask_filled = hcat(mask_filled, falses(ncol(df2)))
             end
-            payload[j][i]  = valuecol[k]
+            i = nrow(df2)
+        else
+            i = refkeycol_pool[refkeycol_refs]
         end
+        if nowarning && mask_filled[j, i]
+            warn("Duplicate entries in unstack at row $k.")
```

Comment:
Would also be nice to print the value and the name of the column.

Reply:
I skip this. It gets messy as I would have to say something like

*duplicate value for index "something which can be multiple columns" and variable "something"*

I feel it is easy enough to to check the contents of the reported column anyway.

Reply:
Sure, you can check it manually, but I find it really makes the difference between software which works and software which is pleasant and efficient to work with. For example, we print the index and the dimensions of the indexed array with `BoundsError`, and it makes a big difference from R where you need to start the debugger to find it out.

Don't do it if that's too much work, but if that's possible it would be really nice, even if the output isn't completely pretty (e.g. it's fine to print a one-element tuple when there's only one column, though `join`'s `last` argument is very useful to format this properly).

Reply:
done!